src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/control/HeaderManager.java

```diff
@@ -57,9 +58,19 @@ public HeaderManager() {
         setProperty(new CollectionProperty(HEADERS, new ArrayList<>()));
     }
 
+    private void assertMutable() {
+        if (isRunningVersion()) {
+            throw new IllegalStateException(
+                    "Cannot modify HeaderManager " + getName() + " while test is running. " +
+                            "If you need dynamic headers, prefer using ${...} expressions in your headers."
```

Comment:
@vlsi probably better to use JMeter functions than "${...} expressions", because in JMeter's manuel the terms is JMeter functions and variables https://jmeter.apache.org/usermanual/functions.html

Reply:
Well, the manual says "Functions and Variables". Do you suggest `If you need dynamic headers, prefer using ${...} functions or variables in your headers`?

Reply:
+1

Reply:
However, I think it makes sense to add a common name for "JMeter expression language"

Reply:
Not sure that the functions in JMeter can be call "expression language", but why not. if you would change to "JMeter expression language", then the manual need to be update (remplace the terms "functions" to "JMeter expression language"), and in JMeter self (menu Tools > Function Helper Dialog)

Reply:
Function Helper Dialog generates functions. I'm not sure it can generate variables.

---

> expression language

Well, the language is something that declares parsing and semantic rules.
For instance, neither function nor variable can specify if nested ${...} are allowed. Neither function nor variable can specify the way to escape $ if the user needs literal $ value without replacements.

Reply:
However, in the context of this failure message "function or variable" sounds good enough